rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,563 @@
+will be done in a phase. Each phase can be gated by a feature flag which is updated to the next phase in the process on
+subsequent Cloud releases. When all the phases have been completed the code can be cleaned up to remove the transitional
+state and feature flags required to get to the target state. The compatibility guarantees of this approach do come with
+a cost though: it will now take longer and require more process to perform the migration. We can however mitigate the
+time required to progress through each phase by increasing the Cloud release cadence.
+
+| Phase | Write                  | Read                                        | Details                                                                       |
+| ----- | ---------------------- | ------------------------------------------- | ----------------------------------------------------------------------------- |
+| 1     | Old key AND new key(s) | Old Key                                     | Introduce new resource(s) and mirror data to new key range on write           |
+| 2     | Old key AND new key(s) | New Key, fallback to old key if not present | Start reading from the new key and using the old key as a fall back           |
+| 3     | Old key AND new key(s) | New Key, fallback to old key if not present | Run data migration that copies all keys from old to new                       |
+| 4     | New key(s)             | New Key                                     | Stop mirroring writes and falling back to old resources when they don't exist |
+| 5     | New key(s)             | New Key                                     | Run optional migration that removes all old keys if they do not have a TTL    |
+
+#### Data Migration Phase
+
+Data migrations no longer need to occur during Auth initialization, which will reduce the time it will take a new Auth
+instance to start processing requests. By waiting until Phase 3 to run data migrations when resource mirroring and
+fallback reads are already being performed by all Auth instances the migration can happen in the background.
+
+This phase is only required for resources which do not heartbeat. For any presence related resources we can use the
+write caused by the next heartbeat to perform the data migration.
+
+> The above only applies when doing phased migrations. Major releases used by self-hosted customers will still be using
+> the one-shot migration approach will still require running the migration during initialization.
+
+### Minor/Additive changes
+
+It's quite common to add a new field to an existing resource in order to support a new feature. All additive fields must
+be optional - adding a field which has a required and meaningful default value causes more issues than just lossy reads
+and writes. Since the versions of instances in the cluster may lag behind Auth considerably, any new fields which need
```

Comment:
Is this saying we should not add a new field that defaults to some new behavior if unset?

For example, we're adding a new `billing_mode` for DyanmoDB and defaulting to on-demand. I know this particular change doesn't ever make it to the backend, so migrations are not a concern, but is this type of change something that fits or does not fit with your recommendations?

Reply:
So, to be clear, the proposal here is that all new fields are always nullable/optional? Or is it okay for new fields to have a default, so long as the default behavior remains the same as it was prior to adding the field?